Crash when calling function in rego

Crash when calling function under `opa run`

# Steps to Reproduce the Problem

Attached below is the entire input to `opa run` I've used. `opa run` has crashed.

# Additional Info

```
$ opa run
OPA 0.10.2 (commit 6af8da83, built at 2018-12-10T22:16:11Z)

Run 'help' to see a list of commands.

> asdf
1 error occurred: 1:1: rego_unsafe_var_error: var asdf is unsafe
> unknown
usage: unknown <unknown-1> [<unknown-2> [...]] (hint: try just 'input')
> input
undefined
> unknown asdf
expected ref but got asdf
> unknown ref
expected ref but got ref
> unknown ref-1
failed to parse ref: expected term but got [minus ref 1]
> unknown ref-asdf
failed to parse ref: expected term but got [minus ref asdf]
> unknown refasdf
expected ref but got refasdf
> hod3
1 error occurred: 1:1: rego_unsafe_var_error: var hod3 is unsafe
> hod3() = {
| asdf
| }
| 
1 error occurred: 1:1: rego_unsafe_var_error: var asdf is unsafe
> hod3(asdf) = {
| asdf
| }
| 
1 error occurred: 1:1: rego_unsafe_var_error: var asdf is unsafe
> hod3(asdf) {
| asdf
| }
| 
Rule 'hod3' defined in package repl. Type 'show' to see rules.
> hod3
1 error occurred: 1:1: rego_type_error: undefined ref: data.repl.hod3
	data.repl.hod3
	^^^^^^^^^^^^^^
	have: any => boolean
> hod3(3)
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0xaf291a]

goroutine 1 [running]:
github.com/open-policy-agent/opa/topdown.newSaveSetElem(0xc42000e018, 0x1, 0x1, 0xc4204ada80, 0xc42000f980)
	/go/src/github.com/open-policy-agent/opa/topdown/save.go:106 +0x8a
github.com/open-policy-agent/opa/topdown.(*saveSet).Push(0xc42000f980, 0xc42000e018, 0x1, 0x1, 0xc4204ada80)
	/go/src/github.com/open-policy-agent/opa/topdown/save.go:29 +0x49
github.com/open-policy-agent/opa/topdown.newSaveSet(0xc42000e018, 0x1, 0x1, 0xc4204ada80, 0x1)
	/go/src/github.com/open-policy-agent/opa/topdown/save.go:24 +0xb9
github.com/open-policy-agent/opa/topdown.(*Query).PartialRun(0xc4204c65b0, 0xeb6580, 0xc4200b4010, 0xc4200b4010, 0xc420032120, 0xc4204ada20, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/github.com/open-policy-agent/opa/topdown/query.go:151 +0x2ce
github.com/open-policy-agent/opa/rego.(*Rego).partial(0xc420001b00, 0xeb6580, 0xc4200b4010, 0xc42000f8f8, 0x1, 0x1, 0xeac480, 0xc420351b60, 0xe30946, 0x7, ...)
	/go/src/github.com/open-policy-agent/opa/rego/rego.go:837 +0x667
github.com/open-policy-agent/opa/rego.(*Rego).Partial(0xc420001b00, 0xeb6580, 0xc4200b4010, 0x0, 0x0, 0x0)
	/go/src/github.com/open-policy-agent/opa/rego/rego.go:499 +0x1af
github.com/open-policy-agent/opa/repl.(*REPL).evalPartial(0xc4202dd200, 0xeb6580, 0xc4200b4010, 0xc42033da70, 0x0, 0x0, 0xc42000f808, 0x1, 0x1, 0xc42000f808, ...)
	/go/src/github.com/open-policy-agent/opa/repl/repl.go:919 +0x3db
github.com/open-policy-agent/opa/repl.(*REPL).evalStatement(0xc4202dd200, 0xeb6580, 0xc4200b4010, 0xde26a0, 0xc42036a480, 0x1, 0x1)
	/go/src/github.com/open-policy-agent/opa/repl/repl.go:822 +0x307
github.com/open-policy-agent/opa/repl.(*REPL).evalBufferOne(0xc4202dd200, 0xeb6580, 0xc4200b4010, 0x0, 0x1)
	/go/src/github.com/open-policy-agent/opa/repl/repl.go:703 +0x2b5
github.com/open-policy-agent/opa/repl.(*REPL).OneShot(0xc4202dd200, 0xeb6580, 0xc4200b4010, 0xc420348ba0, 0x7, 0x0, 0x0)
	/go/src/github.com/open-policy-agent/opa/repl/repl.go:278 +0x38e
github.com/open-policy-agent/opa/repl.(*REPL).Loop(0xc4202dd200, 0xeb6580, 0xc4200b4010)
	/go/src/github.com/open-policy-agent/opa/repl/repl.go:181 +0x303
github.com/open-policy-agent/opa/runtime.(*Runtime).StartREPL(0xc4202437a0, 0xeb6580, 0xc4200b4010)
	/go/src/github.com/open-policy-agent/opa/runtime/runtime.go:295 +0x232
github.com/open-policy-agent/opa/cmd.init.7.func1(0xc420196a00, 0xc4201cc110, 0x1, 0x1)
	/go/src/github.com/open-policy-agent/opa/cmd/run.go:122 +0x4fd
github.com/open-policy-agent/opa/vendor/github.com/spf13/cobra.(*Command).execute(0xc420196a00, 0xc4201cc0e0, 0x1, 0x1, 0xc420196a00, 0xc4201cc0e0)
	/go/src/github.com/open-policy-agent/opa/vendor/github.com/spf13/cobra/command.go:766 +0x2c1
github.com/open-policy-agent/opa/vendor/github.com/spf13/cobra.(*Command).ExecuteC(0x13e04e0, 0xc420075f68, 0xc93823, 0xc420075f78)
	/go/src/github.com/open-policy-agent/opa/vendor/github.com/spf13/cobra/command.go:852 +0x30a
github.com/open-policy-agent/opa/vendor/github.com/spf13/cobra.(*Command).Execute(0x13e04e0, 0x2, 0xc420153f50)
	/go/src/github.com/open-policy-agent/opa/vendor/github.com/spf13/cobra/command.go:800 +0x2b
main.main()
	/go/src/github.com/open-policy-agent/opa/main.go:12 +0x2d
```

--- a/repl/repl.go
+++ b/repl/repl.go
@@ -482,18 +482,27 @@ func (r *REPL) cmdTypes() error {
 	return nil
 }
 
+var errUnknownUsage = fmt.Errorf("usage: unknown <input/data reference> [<input/data reference> [...]] (hint: try 'input')")
+
 func (r *REPL) cmdUnknown(s []string) error {
+
 	if len(s) == 0 && len(r.unknowns) == 0 {
-		return fmt.Errorf("usage: unknown <unknown-1> [<unknown-2> [...]] (hint: try just 'input')")
+		return errUnknownUsage
 	}
-	r.unknowns = make([]*ast.Term, len(s))
-	for i := range r.unknowns {
+
+	unknowns := make([]*ast.Term, len(s))
+
+	for i := range unknowns {
+
 		ref, err := ast.ParseRef(s[i])
 		if err != nil {
-			return err
+			return errUnknownUsage
 		}
-		r.unknowns[i] = ast.NewTerm(ref)
+
+		unknowns[i] = ast.NewTerm(ref)
 	}
+
+	r.unknowns = unknowns
 	return nil
 }
 

--- a/repl/repl_test.go
+++ b/repl/repl_test.go
@@ -549,6 +549,26 @@ func TestUnknownJSON(t *testing.T) {
 	}
 }
 
+func TestUnknownInvalid(t *testing.T) {
+	ctx := context.Background()
+	store := inmem.New()
+	var buffer bytes.Buffer
+	repl := newRepl(store, &buffer)
+
+	err := repl.OneShot(ctx, "unknown x-1")
+	if err == nil || !strings.Contains(err.Error(), "usage: unknown <input/data reference>") {
+		t.Fatal("expected error from setting bad unknown but got:", err)
+	}
+
+	// Ensure that partial evaluation has not been enabled.
+	buffer.Reset()
+	repl.OneShot(ctx, "1+2")
+	result := strings.TrimSpace(buffer.String())
+	if result != "3" {
+		t.Fatal("want true but got:", result)
+	}
+}
+
 func TestUnset(t *testing.T) {
 	ctx := context.Background()
 	store := inmem.New()

